### PR TITLE
NAS-110704 / 21.06 / Do not remove snapshot fields before running annotate_snapshots on them

### PR DESCRIPTION
**Pull request by themylogin**

### CODE REVIEW

File: src/middlewared/middlewared/plugins/zfs.py [hunk 1]
```diff
@@ -1002,12 +1002,20 @@ def query(self, filters, options):
...
-            result = self.middleware.call_sync('zettarepl.annotate_snapshots', [result])[0]
+        if not select or 'retention' in select:
+            if isinstance(result, list):
+                result = self.middleware.call_sync('zettarepl.annotate_snapshots', result)
+            elif isinstance(result, dict):
+                result = self.middleware.call_sync('zettarepl.annotate_snapshots', [result])[0]
+
+        if select:
+            if isinstance(result, list):
+                result = [{k: v for k, v in item.items() if k in select} for item in result]
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/zfs.py:1016]: Why don't we do this before doing `filter_list` ? If a user has thousands of snapshots, we would be iterating twice ?

File: src/middlewared/middlewared/plugins/zfs.py [hunk 2]
```diff
@@ -1002,12 +1002,20 @@ def query(self, filters, options):
...
+            elif isinstance(result, dict):
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/zfs.py:1017]: Should we also be doing this if a snapshot is queried directly ? ( the `id` case )

File: src/middlewared/middlewared/plugins/zfs.py (hunk 1 above)
Comment (themylogin) [reply src/middlewared/middlewared/plugins/zfs.py:1016]: Do what exactly?

File: src/middlewared/middlewared/plugins/zfs.py (hunk 2 above)
Comment (themylogin) [reply src/middlewared/middlewared/plugins/zfs.py:1017]: Sorry, do what?

**Review by sonicaj — APPROVED** [review]

File: src/middlewared/middlewared/plugins/zfs.py (hunk 2 above)
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/zfs.py:1017]: Ah, my bad - the above return is only limited to `name` selection. Thank you for pointing it out

**Pull request merged**